thread-rcu/rcu.h

```diff
@@ -0,0 +1,216 @@
+    } while (0)
+#define barrier() __asm__ __volatile__("" : : : "memory")
+#define smp_mb() __asm__ __volatile__("mfence" : : : "memory")
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#ifdef __CHECKER__
+#define __rcu __attribute__((noderef, address_space(__rcu)))
+#define rcu_check_sparse(p, space) ((void) (((typeof(*p) space *) p) == p))
+#define __force __attribute__((force))
+#define rcu_uncheck(p) ((__typeof__(*p) __force *) p)
+#define rcu_check(p) ((__typeof__(*p) __force __rcu *) p)
+#else
+#define __rcu
+#define rcu_check_sparse(p, space)
+#define __force
+#define rcu_uncheck(p) p
+#define rcu_check(p) p
+#endif /* __CHECKER__ */
+
+/* Avoid false sharing */
+#define __rcu_aligned __attribute__((aligned(128)))
+
+struct rcu_node {
+    unsigned int tid;
+    int rcu_nesting[2];
+    struct rcu_node *next;
+} __rcu_aligned;
+
+struct rcu_data {
+    unsigned int nr_thread;
+    struct rcu_node *head;
+    unsigned int rcu_thrd_nesting_idx;
+    spinlock_t sp;
+};
+
+/* Easy to use */
+#define __rcu_thrd_idx rcu_data.rcu_thrd_nesting_idx
+#define __rcu_thrd_nesting(ptr) \
+    ptr->rcu_nesting[READ_ONCE(__rcu_thrd_idx) & 0x01]
+#define rcu_thrd_nesting __rcu_thrd_nesting(__rcu_per_thrd_ptr)
+
+static struct rcu_data rcu_data = {
+    .nr_thread = 0,
+    .head = NULL,
+    .rcu_thrd_nesting_idx = 0,
+    .sp = SPINLOCK_INIT,
+};
+static __thread struct rcu_node *__rcu_per_thrd_ptr;
+
+static __inline__ struct rcu_node *__rcu_node_add(unsigned int tid)
+{
+    struct rcu_node **indirect = &rcu_data.head;
+    struct rcu_node *node;
+
+    node = (struct rcu_node *) malloc(sizeof(struct rcu_node));
+    if (!node) {
+        fprintf(stderr, "__rcu_node_add: malloc failed\n");
+        abort();
+    }
+
+    node->tid = tid;
+    node->rcu_nesting[0] = 0;
+    node->rcu_nesting[1] = 0;
+    node->next = NULL;
+
+    spin_lock(&rcu_data.sp);
+
+    while (*indirect) {
+        if ((*indirect)->tid == node->tid) {
+            spin_unlock(&rcu_data.sp);
+            free(node);
+            return NULL;
+        }
+        indirect = &(*indirect)->next;
+    }
+
+    *indirect = node;
+    rcu_data.nr_thread++;
+
+    spin_unlock(&rcu_data.sp);
+
+    smp_mb();
+
+    return node;
+}
+
+static __inline__ int rcu_init(void)
+{
+    unsigned int tid = current_tid();
+
+    __rcu_per_thrd_ptr = __rcu_node_add(tid);
+
+    return (__rcu_per_thrd_ptr == NULL) ? -ENOMEM : 0;
+}
+
+static __inline__ void rcu_clean(void)
+{
+    struct rcu_node *node, *tmp;
+
+    spin_lock(&rcu_data.sp);
+
+    for (node = rcu_data.head; node != NULL; node = tmp) {
+        tmp = node->next;
+        free(node);
+    }
+
+    rcu_data.head = NULL;
+    rcu_data.nr_thread = 0;
+
+    spin_unlock(&rcu_data.sp);
+}
+
+/* The per-thread reference count will only modified by their owner
+ * thread but will read by other threads. So here we use WRITE_ONCE().
+ */
+static __inline__ void rcu_read_lock(void)
```

Comment:
Don't use `__inline__`.  Always use `inline` instead.

Reply:
When I wrote this code I use `-std=c89`.
And `inline` is not standard in c89, it is in [c99](https://en.cppreference.com/w/c/language/inline).